launcher/tests/test_notify.py

```diff
@@ -0,0 +1,322 @@
+import datetime
+import os
+import pytest
+import re
+import subprocess
+
+from unittest import mock
+from importlib.machinery import SourceFileLoader
+from tempfile import TemporaryDirectory
+
+relpath_notify = "../sdw_notify/Notify.py"
+path_to_notify = os.path.join(os.path.dirname(os.path.abspath(__file__)), relpath_notify)
+notify = SourceFileLoader("Notify", path_to_notify).load_module()
+
+relpath_updater = "../sdw_updater_gui/Updater.py"
+path_to_updater = os.path.join(os.path.dirname(os.path.abspath(__file__)), relpath_updater)
+updater = SourceFileLoader("Updater", path_to_updater).load_module()
+
+# Regex for warning log if we have no successful update, but uptime is below threshold
+UPTIME_WARNING_REGEX = r"^Uptime \(.* hours\) is above warning threshold \(.* hours\)."
+
+# Regex for info log if we have no successful update, but uptime treshold not reached
+UPTIME_NO_WARNING_REGEX = r"Uptime \(.* hours\) is below warning threshold \(.* hours\)."
+
+# Regex for warning log if we've updated too long ago, and grace period has elapsed
+UPDATER_WARNING_REGEX = r"^Last successful update \(.* hours ago\) is above warning threshold "
+r"\(.* hours\). Uptime grace period of .* hours has elapsed (uptime: .* hours)."
+
+# Regex for info log if we've updated too long ago, but grace period still ticking
+GRACE_PERIOD_REGEX = r"Last successful update \(.* hours ago\) is above "
+r"warning threshold \(.* hours\). Uptime grace period of .* hours has not elapsed "
+r"yet \(uptime: .* hours\)."
+
+# Regex for info log if we've updated recently enough
+NO_WARNING_REGEX = r"Last successful update \(.* hours ago\) is below the warning threshold "
+r"\(.* hours\)."
+
+# Regex for bad contents in `sdw-last-updated` file
+BAD_TIMESTAMP_REGEX = r"Data in .* not in the expected format."
+
+# Regex for lock conflict with updater when launching notifier
+BUSY_LOCK_REGEX = r"Error obtaining lock on '.*'."
+
+# Regex for failure to obtain lock due to permission error
+LOCK_PERMISSION_REGEX = r"Error writing to lock file '.*'"
+
+
+@mock.patch("Notify.sdlog.error")
+@mock.patch("Notify.sdlog.warning")
+@mock.patch("Notify.sdlog.info")
+def test_notify_lock(
+        mocked_info, mocked_warning, mocked_error
+):
+    """
+    Test whether we can successfully obtain an exclusive lock for the notifier
+    script
+    """
+    with TemporaryDirectory() as tmpdir, \
+            mock.patch("Notify.LOCK_FILE_NOTIFIER",
+                       os.path.join(tmpdir, "sdw-notify.lock")):
```

Comment:
FYI: I'm using this pattern to patch the global variable, to avoid unintended side effects either across tests or in a single test; please let me know if there's a preferable pattern for the same result.